refactor: remove "node" type dependency from client runtime

kt3k opened the pull request:
This PR removes 'node.js' type dependency from `runtime/index.d.ts`

There were 2 node.js type dependencies in client runtime.

- `sql-template-tag@4.0.0` depends on Node.js `util`, which was removed in v4.1.0
- `QueryEngineConfig["env"]` and `GetConfigOptions["env"]` use `NodeJS.ProcessEnv` type.

This PR removes the above 2 deps by upgrading `sql-template-tag` and replacing `NodeJS.ProcessEnv` with `Record<string, string | undefined>`

---
*motivations*

This is a preliminary work for Deno support. Deno doesn't have 'node' types. This makes `runtime/index.d.ts` usable as is in deno.
I believe this change is also desirable for other edge runtimes as they don't provide node.js modules by default

kt3k committed:
refactor: remove "node" type dependency from client runtime

Jolg42 commented:
I'll let @aqrln review, note from the past
We mentioned in https://github.com/prisma/prisma/pull/13823 that we want to `We want to upgrade directly to 5.x later.`, I'm not sure why. Maybe it's fine to do the upgrade to `4.1.0` here.

kt3k commented:
`sql-template-tag` dropped `object` type from `RawValue` type (see https://github.com/blakeembrey/sql-template-tag/commit/d1fb92ec5653a5680f0bf9abfbecd4518b56bef4). `RawValue` type is used as parameter of `sql` tagged template literal. That causes type errors in the usages like the below:

```ts
Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${Buffer.from('hello')})`
```
We need to change this, for example, like the below to avoid type error.
```ts
Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${'hello'})`
```

Is this change acceptable? (If not, I'll try `sql-template-tag@5.0.3`, which allow `unknown` type for `RawValue` type.)

kt3k committed:
test: fix type error

kt3k committed:
Revert "test: fix type error"

This reverts commit 97038b7c8996ed548c57e30ffd69a365e158682b.

kt3k commented:
> Prisma.sql`INSERT INTO "Entry" ("id", "binary") VALUES ('4', ${Buffer.from('hello')})`

It looks like these Buffer values actually need to be Buffer type. I'll revert the test changes and try to upgrade `sql-template-tag` to v5.0.3.

kt3k committed:
chore: upgrade sql-template-tag to v5.0.3

kt3k committed:
test: force jest transpile sql-template-tag

kt3k committed:
test: use built version of client runtime in test client

sql-template-tag publishes with ESM only since 5.0.0, and therefore
it can't be 'required' in Node.js env. This commit work around it
by using the built version of runtime, which is in CJS syntax.

kt3k committed:
Merge branch 'main' into refactor-remove-node-deps-from-client-runtime

aqrln commented:
> I'll let @aqrln review, note from the past
We mentioned in https://github.com/prisma/prisma/pull/13823 that We want to upgrade directly to 5.x later., I'm not sure why. Maybe it's fine to do the upgrade to 4.1.0 here.

Sorry for not responding, I was off this Monday and Tuesday.

From the comments and commits it looks like @kt3k has already figured this out (thank you!), but just to confirm and provide additional context: `sql-template-tag@4.1.0` introduced an accidental breaking change that affected us, when we noticed that (which took quite a while because our renovate pipeline was stalled), the author immediately responded to us and released a fix in 5.0.3. However we couldn't upgrade to 5.0.3 because of https://github.com/prisma/prisma/issues/7618.

Here's the relevant PR: https://github.com/prisma/prisma/pull/11506

Adding it to `forceTranspile` like in https://github.com/prisma/prisma/pull/14589/commits/bcc492724c020257dd91ceb5175710034d95f859 might do the trick and make it work, if my understanding is correct (@millsp knows more about it).

aqrln commented on packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -54,6 +54,7 @@ import { getRejectOnNotFound } from './utils/rejectOnNotFound'
 import { serializeRawParameters } from './utils/serializeRawParameters'
 import { validatePrismaClientOptions } from './utils/validatePrismaClientOptions'
 
+const sqltag = sqlTemplateTag.default
```
Maybe `import sqltag, { Sql } from 'sql-template-tag'` instead on line 28 and replace `sqlTemplateTag.Sql` with just `Sql` in the file

kt3k replied:
Good idea! Updated.

aqrln commented on packages/client/src/utils/getTestClient.ts:
```diff
@@ -50,7 +49,7 @@ export async function getTestClient(schemaDir?: string, printWarnings?: boolean)
   const relativeEnvPaths = getEnvPaths(schemaPath, { cwd: absSchemaDir })
   const activeProvider = config.datasources[0].activeProvider
-  const options: GetPrismaClientConfig = {
+  const options = {
```
Was the type annotation removed just because it wasn't necessary or because it was causing some problems?

kt3k replied:
I changed the reference to runtime module from `../runtime/` (`client/src/runtime`) to `../../runtime/` (`client/runtime`) in the above lines, and `client/runtime` doesn't export `GetPrismaClientConfig`. So I removed `GetPrismaClientConfig`. (also it doesn't look absolutely necessary)

I changed the reference to runtime like this because if we import from `client/src/runtime/` it throws require error at `sql-template-tag` because of ESM issue. The built version of runtime is in CommonJS, so it can avoid that issue. I'm not 100% sure this approach is acceptable or not. What do you think?

kt3k replied:
Ah, this change seems causing the mismatch of query engine version in CI

I now see these errors..
```
Query engine library for current platform "debian-openssl-1.1.x" could not be found.
    You incorrectly pinned it to debian-openssl-1.1.x

    This probably happens, because you built Prisma Client on a different platform.
    (Prisma Client looked in "/home/runner/work/prisma/prisma/packages/client/runtime/libquery_engine-debian-openssl-1.1.x.so.node")
```

millsp replied:
It could have been a viable option to call `generateInFolder` instead, that might have solved the issue. I feel like using the runtime as it is shipped (like you proposed) is a viable option. The tests in `__tests__` are currently being upgraded to a cleaner and more standard setup, so it is only a matter of time until we remove them. Additionally, swapping `getPrismaClient` to `generateInFolder` should be transparent.

millsp replied:
(see https://github.com/prisma/prisma/pull/14589#discussion_r947135925)

kt3k committed:
refactor: simplify import statement

kt3k commented:
@aqrln Thank you for the feedbacks!

> However we couldn't upgrade to v5 because of https://github.com/prisma/prisma/issues/7618.

This problem seems very complex to me. I've been trying a few workarounds for this, but I still see the errors..

> Adding it to forceTranspile like in https://github.com/prisma/prisma/commit/bcc492724c020257dd91ceb5175710034d95f859 might do the trick and make it work, if my understanding is correct, but I'm not sure (@millsp knows more about it).

This seems to solve the issue for the case when sql-template-tag is directly imported in jest environment.

But there are other cases where test case invokes another node.js process (e.g. `packages/client/src/__tests__/integration/happy/signals/test.ts`), and that process imports sql-template-tag, which still causes error while importing it.

To avoid that one, I did [this change](https://github.com/prisma/prisma/pull/14589/commits/92277d445893b89a6ba3f5dc77fff33f4e5db0f3), and now I see another error about version mismatch of query engines..

kt3k commented:
In what path does the query engine exist in CI?

The error message says it couldn't find it in the following paths, for example:
```
/home/ubuntu/actions-runner/_work/prisma/prisma/.prisma/client
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/runtime
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
/tmp/prisma-engines
/home/ubuntu/actions-runner/_work/prisma/prisma/packages/client/src/__tests__/integration/happy/insensitive-postgresql-feature-flag
```
But on my local machine, the query engine exists at `packages/client/libquery_engine-darwin.dylib.node`. So it feels strange that it couldn't find it from `packages/client`..

millsp commented:
> In what path does the query engine exist in CI?

The engine is copied in two locations, directly into `@prisma/engines` on postinstall, which has also the side-effect to put a copy into the cache (if possible in eg. `/home/.cache/prisma/...`. Then there is a third location for our new `packages/client/tests/functional` (which is an effort to port tests from `__tests__/integration`), in this case the engine is also copied to the root of `packages/client`. Which tests are failing?

kt3k commented:
Thanks for the explanation!

> Which tests are failing?

Some examples of failing tests are:

- `packages/client/src/__tests__/integration/happy/rejectNotFound/test.ts`
- `packages/integration-tests/src/__tests__/integration/sqlite/runtime.test.ts`
- `packages/integration-tests/src/__tests__/integration/postgresql/runtime.test.ts`
- `packages/integration-tests/src/__tests__/integration/mariadb/runtime.test.ts`

kt3k commented:
Having looked at details of current test failures, the number of failures are overwhelming. I think it's not a good idea to to do workaround for `sql-template-tag@5.0.3`. Especially this commit [92277d4](https://github.com/prisma/prisma/pull/14589/commits/92277d445893b89a6ba3f5dc77fff33f4e5db0f3) looks bad

I'd rather try type casting with `sql-template-tag@4.1.0`, as the problem of it is only about typing

aqrln commented:
> I'd rather try type casting with sql-template-tag@4.1.0, as the problem of it is only about typing

The problem is we're re-exporting the template tag from that library. We could work around it for ourselves with type casting, but our users' code would still be broken with that version.

kt3k committed:
Revert "refactor: simplify import statement"

This reverts commit a30eb1286c6eeae480ac572f5803195e05485d8b.

kt3k committed:
Revert "test: use built version of client runtime in test client"

This reverts commit 92277d445893b89a6ba3f5dc77fff33f4e5db0f3.

kt3k committed:
Revert "test: force jest transpile sql-template-tag"

This reverts commit bcc492724c020257dd91ceb5175710034d95f859.

kt3k committed:
Revert "chore: upgrade sql-template-tag to v5.0.3"

This reverts commit 8938c23a73a80eeb2f5e53468a021149cffa8ebb.

kt3k committed:
Merge branch 'main' into refactor-remove-node-deps-from-client-runtime

kt3k committed:
fix: work around type bug of sql-template-tag@4.1.0

kt3k commented:
@millsp @aqrln I think this PR is now ready for review. PTAL

millsp commented on packages/client/src/runtime/utils/sql-template-tag.ts:
```diff
@@ -0,0 +1,16 @@
+declare function sqltagFunc(strings: ReadonlyArray<string>, ...values: RawValue[]): Sql
+
+const join = joinOriginal as typeof joinFunc
+const sqltag = sqltagOriginal as typeof sqltagFunc
```
This looks good although I prefer the alternative with `generateInFolder`, if you have time and if it actually works out that easily. This seems to be viable, but of course it adds that thin layer between us and the library, which I'd like to avoid if possible. If not, `__tests__` are being phased out, and we can add a TODO as a comment here that when we don't depend on `getTestClient` anymore, we can basically upgrade to v5 since we bundle to CJS.

kt3k replied:
Ok, I'll experiment with `generateInFolder` for a bit.

millsp replied:
If it is too tricky to get the integration-tests to pass, just leave it as is and ping me.

kt3k committed:
test: use generateTestClient instead of getTestClient

kt3k committed:
Merge branch 'main' into refactor-2

kt3k committed:
Merge branch 'main' into refactor-2

kt3k commented on packages/integration-tests/src/__tests__/__helpers__/integrationTest.ts:
```diff
@@ -239,7 +239,13 @@ export function runtimeIntegrationTest<Client>(input: Input<Client>) {
       states[scenario.name] = state
 
-      const PrismaClient = await getTestClient(ctx.fs.cwd())
+      await generateInFolder({
```
Changed this `getTestClient` call to `generateInFolder`. This has currently 2 problems:

- The test fails with OOM error (if I reduces the number of scenarios, then they pass.)
- The assertions about Decimal don't pass because the Decimal constructors don't match

millsp replied:
I see, let's keep your original work, I hoped this would not get too tricky. Thanks for trying!

millsp replied:
(these tests are also legacy tests that we will eventually get rid of)

kt3k replied:
Ok, I'll move this back to `sql-template-tag@4.1.0 + type modification` version

kt3k committed:
Revert "test: use generateTestClient instead of getTestClient"

This reverts commit 37e1563b29f3c2ab449b2664d40cad238013fc30.

kt3k committed:
docs: add todo comment for sql-template-tag wrapper script

millsp commented on packages/client/src/runtime/utils/sql-template-tag.ts:
```diff
@@ -0,0 +1,22 @@
+// TODO: We should remove this file when we upgrade sql-template-tag to 5.0.3 or above.
+// Note: The reason why we don't upgrade it now is that it publishes with ESM syntax.
+// We can upgrade it when we get rid of the use of `getTestClient` in testing.
+// See the below for more context:
```
Awesome, thank you.

kt3k committed:
Merge branch 'main' into refactor-2

aqrln reviewed (approved):
Nice work, thank you!

kt3k committed:
Merge branch 'main' into refactor-2

millsp commented:
Hey @kt3k, I have taken some time to upgrade to `sql-template-tag@5` via this [PR](https://github.com/prisma/prisma/pull/14885). I'm waiting for reviews on it, yours is also welcome :)

kt3k commented:
@millsp Great! Thank you!

I'll update this PR soon.

kt3k committed:
Merge branch 'main' into refactor-2

kt3k committed:
chore: remove unused workaround

kt3k committed:
Merge branch 'main' into refactor-2

kt3k committed:
chore: revert irrelevant change

kt3k committed:
Merge branch 'main' into refactor-remove-node-deps-from-client-runtime

kt3k commented:
Updated. Now this PR only has `NodeJS.ProcessEnv | Record<string, string>` -> `Record<string, string | undefined>` update and the test of the runtime declaration file.

aqrln reviewed (approved):
LGTM!

millsp reviewed (approved):
Thank you for this PR and your efforts to get Prisma working with Deno :sauropod: :rocket:
